kong/router/atc_compat.lua

```diff
@@ -0,0 +1,468 @@
+    services = {},
+    fields = nil,
+  }, _MT)
+
+  for _, r in ipairs(routes) do
+    router.routes[r.route.id] = r.route
+    router.services[r.route.id] = r.service
+
+    get_atc(r.route)
+    assert(inst:add_matcher(route_priority(r.route), r.route.id, get_atc(r.route)))
+
+    router.fields = inst:get_fields()
+  end
+
+  return router
+end
+
+
+function _M:select(req_method, req_uri, req_host, req_scheme,
+                   _src_ip, _src_port,
+                   _dst_ip, _dst_port,
+                   sni, req_headers)
+  if req_method and type(req_method) ~= "string" then
+    error("method must be a string", 2)
+  end
+  if req_uri and type(req_uri) ~= "string" then
+    error("uri must be a string", 2)
+  end
+  if req_host and type(req_host) ~= "string" then
+    error("host must be a string", 2)
+  end
+  if req_scheme and type(req_scheme) ~= "string" then
+    error("scheme must be a string", 2)
+  end
+  if src_ip and type(src_ip) ~= "string" then
+    error("src_ip must be a string", 2)
+  end
+  if src_port and type(src_port) ~= "number" then
+    error("src_port must be a number", 2)
+  end
+  if dst_ip and type(dst_ip) ~= "string" then
+    error("dst_ip must be a string", 2)
+  end
+  if dst_port and type(dst_port) ~= "number" then
+    error("dst_port must be a number", 2)
+  end
+  if sni and type(sni) ~= "string" then
+    error("sni must be a string", 2)
+  end
+  if req_headers and type(req_headers) ~= "table" then
+    error("headers must be a table", 2)
+  end
+
+  local c = context.new(self.schema)
+
+  for _, field in ipairs(self.fields) do
+    if field == "http.method" then
+      assert(c:add_value("http.method", req_method))
+
+    elseif field == "http.path" then
+      assert(c:add_value("http.path", req_uri))
+
+    elseif field == "http.host" then
+      assert(c:add_value("http.host", req_host))
+
+    elseif field == "net.protocol" then
+      assert(c:add_value("net.protocol", req_scheme))
+
+    elseif field == "tls.sni" then
+      assert(c:add_value("tls.sni", sni))
+
+    elseif req_headers and field:sub(1, 13) == "http.headers." then
+      local h = field:sub(14)
+      local v = req_headers[h]
+
+      if v then
+        if type(v) == "string" then
+          assert(c:add_value(field, v:lower()))
+
+        else
+          -- TODO: support array of values
+          assert(c:add_value(field, v[1]:lower()))
+        end
+      end
+    end
+  end
+
+  local matched = self.router:execute(c)
+  if not matched then
+    return nil
+  end
+
+  local uuid, matched_path, captures = c:get_result()
+
+  return {
+    route           = self.routes[uuid],
+    service         = self.services[uuid],
+    prefix          = matched_path,
+    matches = {
+      uri_captures = (captures and captures[1]) and captures or nil,
+    }
+  }
+end
+
+
+function _M:exec(ctx)
+  local req_method = get_method()
+  local req_uri = ctx and ctx.request_uri or var.request_uri
+  local req_host = var.http_host
```

Comment:
It seems there is no define of `var`, should we add `ngx.var` at the top of file?